CAIPs/caip-x.md

```diff
@@ -0,0 +1,339 @@
+
+## Motivation
+
+The motivation behind this proposal is to enhance the flexibility of CAIP-25 by enabling management of session permissions without sessionIds which don't map well to extension based wallet's dapp connections and could therefore add constraints and burdens to existing flows. The proposed methods provide an intuitive way to add, revoke, and retrieve permissions within an existing session, simplifying the management of session lifecycles.
+
+### Use Case Scenarios
+
+1. **Dapp Initiated Adding Permissions To an Existing Session:**
+
+   - **Current Method:** Call `provider_authorize` again with an existing session identifier and new scopes/methods to add. This is actually somewhat ambiguous in CAIP-25 where it’s unclear if incremental permissions adds should include the full object or just scopes to add: "This object gets updated, extended, closed, etc. by successive calls and notifications, each tagged by this identifier."
+   - **Proposed Method:** Use `provider_augment` to add new permissions to the existing session. Alternatively, this could be achieved with more specific language indicating that a subsequent `provider_authorize` call with new scopes/permissions could simply add to an existing session. If we pursue this latter approach one flow that will need clarity is what happens if a subsequent `provider_authorize` request includes `requiredScopes` which the respondent does not authorize. In this case I would expect that this does not revoke the existing session but rather only the new request is rejected.
+
+2. **Wallet Initiated Adding Permissions To an Existing Session:**
+
+   - **Current Method:** CAIP-25 does not make it very clear how a respondent (wallet), can modify the permissions of an existing session. The following excerpt is the closest we get: "The properties and authorization scopes that make up the session are expected to be persisted and tracked over time by both parties in a discrete data store, identified by an entropic identifier assigned in the initial response. This object gets updated, extended, closed, etc. by successive calls and notifications, each tagged by this identifier."
+   - **Proposed Method:** Wallet publishes and caller/dapp listens for an event `providerAuthorizationChanged` with the new full sessionScope.
+
+3. **Wallet Initiated Permissions Revocation:**
+
+   - **Current Method:** "If a respondent (e.g. a wallet) needs to initiate a new session, whether due to user input, security policy, or session expiry reasons, it can simply generate a new session identifier to signal this notification to the calling provider."
+   - Given this language it is unclear if a wallet can revoke permissions without creating a new session. It is also therefore unclear if a wallet can revoke some subset of permissions without creating a new session.
+   - **Proposed Method:** Wallet publishes and caller/dapp listens for an event `providerAuthorizationChanged` with the new full sessionScope.
```

Comment:
I'm assuming you're referring to user-initiated permission revocation in the extension? If so, what happens if the dapp is not connected to the wallet when the permission is revoked? Can the dapp just not rely on `provider_getSession`?

Reply:
> I'm assuming you're referring to user-initiated permission revocation in the extension? If so, what happens if the dapp is not connected to the wallet when the permission is revoked?

🤔 I guess you mean if the connection is severed somehow? I suppose in that case the dapp would use `provider_getSession` but I think this event is useful for most cases where a dapp can immediately reflect a change to the state of its permissions

Reply:
severed, or just timed out.  i think the fundamental assumption that both app and wallet persist the config (if not the "session" in browser-memory or RPC-connection sense) is the first thing we need to be aligned on, before getting to other stuff?  in hassan's example, the wallet would let the user revoke a permission and... persist that message/event until there's an active RPC connection, right?

all this sounds fine to me, but should be explicit in the spec since it obligates the provider (or the wallet+plus+provider) to persist until next connection.  conversely, if this event is ONLY fired within an active session and dropped otherwise, specify that and separately define the behavior if user tells wallet to revoke without (or without knowing about the absence of) a live connection

Reply:
Does the following additional sentence assuage your concerns?

```
If a connection between the wallet and the caller/dapp is severed and the possiblity of missed events arises, the caller/dapp should immediately call `wallet_getSession` to retrieve the current session scopes.
```